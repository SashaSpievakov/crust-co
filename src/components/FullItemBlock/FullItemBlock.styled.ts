import styled from 'styled-components';
import { Container } from '../../styles/Base.styled';

const Item = styled(Container)`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;

  @media screen and (max-width: 600px) {
    gap: 15px;
  }

  @media screen and (max-width: 400px) {
    gap: 10px;
  }
`;

const Image = styled.img`
  margin-top: 60px;
  margin-bottom: 40px;
  border-radius: 20px;

  @media screen and (max-width: 600px) {
    width: 300px;
    height: 300px;
    margin-top: 50px;
    margin-bottom: 10px;
  }

  @media screen and (max-width: 400px) {
    width: 200px;
    height: 200px;
    margin-top: 20px;
    margin-bottom: 0;
  }
`;

const Title = styled.h2`
  font-size: 45px;
  font-weight: 700;
  text-align: center;

  @media screen and (max-width: 600px) {
    font-size: 35px;
    margin-bottom: 0;
  }

  @media screen and (max-width: 400px) {
    line-height: 35px;
  }
`;

const Paragraph = styled.p`
  font-size: 19px;
  max-width: 600px;
  text-align: center;
  margin-bottom: 15px;

  @media screen and (max-width: 550px) {
    font-size: 14px;
  }
`;

const Price = styled.h3`
  font-size: 45px;

  @media screen and (max-width: 500px) {
    font-size: 40px;
  }
`;

export { Item, Image, Title, Paragraph, Price };
