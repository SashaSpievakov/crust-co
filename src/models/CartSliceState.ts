import { ICartItem } from './ICartItem';

export interface CartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
