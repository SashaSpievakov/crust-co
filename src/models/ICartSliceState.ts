import { ICartItem } from './ICartItem';

export interface ICartSliceState {
  totalPrice: number;
  itemsCount: number;
  items: ICartItem[];
}
