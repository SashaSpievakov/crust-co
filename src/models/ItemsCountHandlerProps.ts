export interface ItemsCountHandlerProps {
  id: string;
  name: string;
  price: number;
  activeSize: number;
  activeType: number;
  typeNames: string[];
  isFullScreen?: boolean;
}
