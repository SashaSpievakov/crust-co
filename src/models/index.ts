export type { ICartItem } from './ICartItem';
export type { IPizzaItem } from './IPizzaItem';
