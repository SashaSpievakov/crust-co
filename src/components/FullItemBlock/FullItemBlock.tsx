import { useState } from 'react';

import { IPizzaItem } from '../../models';
import ItemsCountHandler from '../ItemsCountHandler/ItemsCountHandler';
import Selector from '../Selector/Selector';
import { ButtonLink } from '../UI';
import { Image, Item, Paragraph, Price, Title } from './FullItemBlock.styled';

interface FullItemProps {
  item: IPizzaItem;
}

export const typeNames: string[] = ['traditional', 'thin'];

const FullItemBlock = ({ item }: FullItemProps) => {
  const [activePrice, setActivePrice] = useState<number>(item.price);
  const [activeSize, setActiveSize] = useState<number>(12);
  const [activeType, setActiveType] = useState<number>(0);

  return (
    <Item data-testid="itemPage">
      <picture>
        <source
          srcSet={`../assets/img/pizza${item.id}.webp`}
          type="image/webp"
        />
        <source srcSet={`../assets/img/pizza${item.id}.png`} type="image/png" />
        <Image
          src={`../assets/img/pizza${item.id}.png`}
          alt={`${item.name} pizza`}
          width="450"
          height="450"
        />
      </picture>
      <Title>{item.name}</Title>
      <Paragraph>{item.description}</Paragraph>
      <Selector
        price={activePrice}
        sizes={item.sizes}
        types={item.types}
        activeSize={activeSize}
        activeType={activeType}
        setActivePrice={setActivePrice}
        setActiveSize={setActiveSize}
        setActiveType={setActiveType}
        typeNames={typeNames}
        isFullScreen
      />
      <Price>{activePrice}$</Price>
      <ItemsCountHandler
        id={item.id}
        name={item.name}
        price={activePrice}
        activeSize={activeSize}
        activeType={activeType}
        typeNames={typeNames}
        isFullScreen
      />
      <ButtonLink link="/">Go Back</ButtonLink>
    </Item>
  );
};

export default FullItemBlock;
