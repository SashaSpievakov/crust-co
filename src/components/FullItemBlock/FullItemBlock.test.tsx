import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import FullItemBlock from './FullItemBlock';
import rendererWithProviders from '../../tests/helpers/rendererWithProviders';
import renderWithProviders from '../../tests/helpers/renderWithProviders';
import { IPizzaItem } from '../../models/IPizzaItem';

const FullItemProp: IPizzaItem = {
  id: '12',
  name: 'Chicken Curry',
  description:
    'Lorem ipsum dolor sit amet consectetur, adipisicing elit. Sed vitae, ab eum nam facilis, soluta corporis molestiae quam, quis cumque consequuntur suscipit dolore praesentium fugiat minima voluptas doloribus aliquid optio!',
  types: [0, 1],
  sizes: [12, 14, 16],
  price: 14,
  category: 1,
  rating: 11,
};

describe('FullItemBlock Test', () => {
  test('renders the FullItemBlock component', () => {
    const snapshot = rendererWithProviders(
      <FullItemBlock item={FullItemProp} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  test('checks link redirect to the Home page', () => {
    renderWithProviders(<FullItemBlock item={FullItemProp} />, true, '/cart');
    const links = screen.getAllByRole('link', {
      name: /go back/i,
    });

    userEvent.click(links[1]);

    expect(
      screen.getByRole('heading', {
        name: /all pizzas/i,
      }),
    ).toBeInTheDocument();
  });

  describe('checks the price changing', () => {
    beforeEach(() => {
      renderWithProviders(<FullItemBlock item={FullItemProp} />);
    });

    test('checks type click', () => {
      const typeItem = screen.getByText(/thin/i);
      const price = screen.getByRole('heading', {
        level: 4,
      });

      userEvent.click(typeItem);

      expect(price).toHaveTextContent('15$');
    });

    test('checks size click', () => {
      const sizeItem = screen.getByText(/16 inch/i);
      const price = screen.getByRole('heading', {
        level: 4,
      });

      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('18$');
    });

    test('checks type and size click', () => {
      const typeItem = screen.getByText(/thin/i);
      const sizeItem = screen.getByText(/14 inch/i);
      const price = screen.getByRole('heading', {
        level: 4,
      });

      userEvent.click(typeItem);
      userEvent.click(sizeItem);

      expect(price).toHaveTextContent('17$');
    });
  });
});
