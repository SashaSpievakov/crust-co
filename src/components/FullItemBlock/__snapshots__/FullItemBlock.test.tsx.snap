// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullItemBlock Tests renders the FullItemBlock component 1`] = `
Object {
  "children": Array [
    <img
      alt="pizza"
      className="sc-dkrFOg fXUcwl"
      height="450"
      src="../assets/img/pizza9.png"
      width="450"
    />,
    <h2
      className="sc-hLBbgP fiXyOf"
    >
      Chicken Curry
    </h2>,
    <p
      className="sc-eDvSVe jzLwQf"
    >
      Lorem, ipsum dolor sit amet consectetur adipisicing elit. Aliquid nisi aspernatur debitis quod suscipit, vitae placeat natus ipsum inventore esse rerum animi facere numquam saepe vero mollitia quibusdam at voluptates est commodi laudantium? Iste inventore quibusdam cupiditate nemo sint iusto minus nihil culpa! Sequi architecto nesciunt explicabo mollitia. Laboriosam, odio.
    </p>,
    <section
      className="sc-eDWCr hJjroC"
    >
      <ul>
        <li
          className="sc-bqWxrE bUZLVc"
          onClick={[Function]}
        >
          traditional
        </li>
        <li
          className="sc-bqWxrE dvnTkm"
          onClick={[Function]}
        >
          thin
        </li>
      </ul>
      <ul>
        <li
          className="sc-bqWxrE dvnTkm"
          onClick={[Function]}
        >
          12
           inch
        </li>
        <li
          className="sc-bqWxrE dvnTkm"
          onClick={[Function]}
        >
          14
           inch
        </li>
        <li
          className="sc-bqWxrE dvnTkm"
          onClick={[Function]}
        >
          16
           inch
        </li>
      </ul>
    </section>,
    <h4
      className="sc-jSUZER glJRmG"
    >
      14
      $
    </h4>,
    <button
      className="sc-gKPRtg sc-jrcTuL sc-ipEyDJ iLACsd kKLFCc heSXxL"
      onClick={[Function]}
    >
      <svg
        className="sc-bjfHbI lhpcBk"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
        />
      </svg>
      <span>
        Add
      </span>
    </button>,
    <a
      className="sc-gKPRtg sc-iBYQkv iLACsd gcostn"
      href="/"
      onClick={[Function]}
    >
      <span>
        Go Back
      </span>
    </a>,
  ],
  "props": Object {
    "className": "sc-bcXHqe sc-gswNZR bZXhLi edAeMk",
  },
  "type": "div",
}
`;
